python/ray/data/_internal/execution/interfaces/execution_options.py

```diff
@@ -192,6 +192,28 @@ def satisfies_limit(self, limit: "ExecutionResources") -> bool:
             and self.object_store_memory <= limit.object_store_memory
         )
 
+    def satisfies_limit_with_reason(self, limit: "ExecutionResources"):
```

Comment:
let's not diverge the code path. We can just make the original `satisfies_limit` method return which resources are not satisfied. And returning a Tuple[bool, list] would be more clear than returning different types.